app/src/pages/api/stripe/webhook.ts

```diff
@@ -0,0 +1,124 @@
+import { type NextApiRequest, type NextApiResponse } from "next";
+import Stripe from "stripe";
+import { prisma } from "~/server/db";
+import { env } from "~/env.mjs";
+import { emailAdminsAboutPaymentFailure } from "~/server/utils/emails";
+
+const stripe = new Stripe(env.STRIPE_SECRET_KEY ?? "");
+const endpointSecret = env.STRIPE_WEBHOOK_SECRET ?? "";
+
+// Stripe needs raw body
+export const config = {
+  api: {
+    bodyParser: false,
+  },
+};
+
+const readRequestBody = (req: NextApiRequest): Promise<Buffer> => {
+  return new Promise((resolve, reject) => {
+    let data = "";
+    req.on("data", (chunk) => (data += chunk));
+    req.on("end", () => resolve(Buffer.from(data)));
+    req.on("error", (err) => reject(err));
+  });
+};
+
+const handler = async (req: NextApiRequest, res: NextApiResponse) => {
+  if (req.method !== "POST") {
+    res.setHeader("Allow", "POST");
+    return res.status(405).end("Method Not Allowed");
+  }
+
+  const rawBody = await readRequestBody(req);
+  const sig = req.headers["stripe-signature"];
+
+  let event;
+  try {
+    event = stripe.webhooks.constructEvent(rawBody, sig ?? "", endpointSecret);
+  } catch {
+    return res.status(400).send(`Webhook Error`);
+  }
+
+  switch (event.type) {
+    case "charge.succeeded":
+      const paymentIntent = event.data.object;
+      const invoiceId = paymentIntent.metadata.invoiceId;
```

Comment:
This seems like a good place to check if the invoiceId exists. That way we can change the function signature of all the functions you call later to expect that invoiceId is a string, not string or undefined.

Reply:
For some reason, I could not do it. returning and breaking the switch/case did not help with typings. Let me take a look again.

Reply:
Ah, I think I see the problem. You need to define invoiceId before the switch statement instead of in one of its conditional blocks.